### PR TITLE
Add 9.3 in order to clone 9.3 databases to AWS.

### PR DESCRIPTION
**Pull request by alexeyklyukin**
Add yet unreleased changes to wale_restore in order to avoid the
issue with broken xlog symlink.

### CODE REVIEW

File: postgres-appliance/wale_restore.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+                return float(f.read())
+        except Exception:
+            logger.exception('Failed to read PG_VERSION from %s', data_dir)
+    return 0.0
+
+
+class WALERestore(object):
+
+    def __init__(self, scope, datadir, connstring, env_dir, threshold_mb, threshold_pct, use_iam, no_master):
+        self.scope = scope
```
Comment (avaczi) [inline postgres-appliance/wale_restore.py]: How about matching the argument names and the class variables?  Most of them are already called the same...

File: postgres-appliance/wale_restore.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+    def should_use_s3_to_create_replica(self):
+        """ determine whether it makes sense to use S3 and not pg_basebackup """
+
+        threshold_megabytes = self.wal_e.threshold_mb
+        threshold_backup_size_percentage = self.wal_e.threshold_pct
+
+        try:
+            latest_backup = subprocess.check_output(self.wal_e.cmd.split() + ['backup-list', '--detail', 'LATEST'])
+            # name    last_modified   expanded_size_bytes wal_segment_backup_start    wal_segment_offset_backup_start
```
Comment (avaczi) [inline postgres-appliance/wale_restore.py]: in this exceptional case I wouldn't be against having 120+ char rows, to see the structure better (if it's important)

File: postgres-appliance/wale_restore.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+            except psycopg2.Error as e:
+                logger.error('could not determine difference with the master location: %s', e)
+                return False
+        else:
+            # always try to use WAL-E if base backup is available
+            diff_in_bytes = 0
+
+        # if the size of the accumulated WAL segments is more than a certan percentage of the backup size
+        # or exceeds the pre-determined size - pg_basebackup is chosen instead.
+        return (diff_in_bytes < long(threshold_megabytes) * 1048576) and\
```
Comment (avaczi) [inline postgres-appliance/wale_restore.py]: is the backslash needed?

File: postgres-appliance/wale_restore.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        if (ret == 0 and not
+           self.fix_subdirectory_path_if_broken('pg_xlog' if get_major_version(self.data_dir) < 10.0 else 'pg_wal')):
+            return 1
+        return ret
+
+
+def main():
+    logging.basicConfig(format='%(asctime)s %(levelname)s: %(message)s', level=logging.INFO)
+    parser = argparse.ArgumentParser(description='Script to image replicas using WAL-E')
+    parser.add_argument('--scope', required=True)
```
Comment (avaczi) [inline postgres-appliance/wale_restore.py]: what does `scope` actualy do?  it does not seem to be used anywhere...

File: postgres-appliance/wale_restore.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    parser.add_argument('--role', required=False)
```
Comment (avaczi) [inline postgres-appliance/wale_restore.py]: ... and `role` even less

File: postgres-appliance/wale_restore.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    parser.add_argument('--retries', type=int, default=1)
+    parser.add_argument('--envdir', required=True)
+    parser.add_argument('--threshold_megabytes', type=int, default=10240)
+    parser.add_argument('--threshold_backup_size_percentage', type=int, default=30)
+    parser.add_argument('--use_iam', type=int, default=0)
+    parser.add_argument('--no_master', type=int, default=0)
+    args = parser.parse_args()
+
+    # retry cloning in a loop
+    for _ in range(0, args.retries + 1):
```
Comment (avaczi) [inline postgres-appliance/wale_restore.py]: shouldn't we add some sleep between retries?

**Comment by alexeyklyukin** [comment]
The wal_restore.py is a copy of the same file from Patroni.

**Comment by avaczi** [comment]
:+1:

File: postgres-appliance/wale_restore.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+# theoretically should work with SWIFT, but not tested on it
+
+# arguments are:
+#   - cluster scope
+#   - cluster role
+#   - master connection string
+#   - number of retries
+#   - envdir for the WALE env
+# - WALE_BACKUP_THRESHOLD_MEGABYTES if WAL amount is above that - use pg_basebackup
+# - WALE_BACKUP_THRESHOLD_PERCENTAGE if WAL size exceeds a certain percentage of the
```
Comment (a1exsh) [inline postgres-appliance/wale_restore.py]: ...of the?

File: postgres-appliance/wale_restore.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+        self.master_connection = connstring
+        self.data_dir = datadir
+        self.wal_e = namedtuple('wale', 'dir,threshold_mb,threshold_pct,iam_string,cmd')
+        self.wal_e.dir = env_dir
+        self.wal_e.threshold_mb = threshold_mb
+        self.wal_e.threshold_pct = threshold_pct
+        self.wal_e.iam_string = ' --aws-instance-profile ' if use_iam == 1 else ''
+        self.no_master = no_master
+        self.wal_e.cmd = 'envdir {0} wal-e {1} '.format(self.wal_e.dir, self.wal_e.iam_string)
+        self.init_error = (not os.path.exists(self.wal_e.dir))
```
Comment (a1exsh) [inline postgres-appliance/wale_restore.py]: Why not `not(expr)`?

File: postgres-appliance/wale_restore.py (hunk 7 above)
Comment (alexeyklyukin) [inline postgres-appliance/wale_restore.py]: The wal_restore.py is a copy of the same file from Patroni.

File: postgres-appliance/wale_restore.py (hunk 8 above)
Comment (alexeyklyukin) [inline postgres-appliance/wale_restore.py]: The wal_restore.py is a copy of the same file from Patroni.

**Comment by alexeyklyukin** [comment]
I think it makes sense to wait until the next version of Patroni is released, since we merged more changes to wale_restore. I will close this one.

**Pull request closed without merging**